Lastpass lookup plugin

azenk opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### ANSIBLE VERSION

```
ansible 2.2.0 (lpass_lookup 106ec7db0b) last updated 2016/06/16 19:35:29 (GMT -500)
  lib/ansible/modules/core: (detached HEAD d4eaeac8c9) last updated 2016/06/16 14:39:31 (GMT -500)
  lib/ansible/modules/extras: (detached HEAD eed9d601b5) last updated 2016/06/16 14:39:32 (GMT -500)
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

This plugin can be used with the lpass cli interface for lastpass.
[lastpass-cli](https://github.com/lastpass/lastpass-cli)

Example:
Add a lookup to your playbooks/variables somewhere:

```
some_variable: "{{ lookup('lpass','Some Lastpass entry name or ID', field='username') }}"
```

Start a lpass session prior to using ansible, run ansible, logout when finished:

```
lpass login user@domain.com
ansible-playbook foo.yml
lpass logout
```

azenk committed:
Lastpass lookup plugin

This plugin can be used with the lpass cli interface for lastpass.
[lastpass-cli](https://github.com/lastpass/lastpass-cli)

Example:
Add a lookup to your playbooks/variables somewhere:
```
some_variable: "{{ lookup('lastpass','Some Lastpass entry name or ID', field='username') }}"
```
Usage:
* start a lpass session prior to using ansible
* run ansible
* logout when finished

```
lpass login user@domain.com
ansible-playbook foo.yml
lpass logout
```

jhawkesworth commented:
Interesting idea. I wonder if it would be possible/uaeful to implement as a callback plugin.

azenk commented:
I'm not quite sure what the potential benefits to implementing this as a callback plugin rather than a lookup plugin would be.  Can you expand on the thought?

jhawkesworth commented:
Probably not many, on reflection, it was just an idle though.  

I created a callback plugin to acquire kerberos credentials for passwordless connection to windows hosts and just thought maybe the same thing could be done for lastpass logins.  But I guess the difference is you might well have lots of lastpass logins for different hosts/services, whereas on windows you typically have a domain login that lets you do whatever you are allowed to do on lots of machines.

bcoca commented:
a lookup makes more sense as you can do: 

```
ansible_password: "{{lookup('lastpass' ....
```

the kerberos case is 'special' as it was not really a password but a ticket/token you maintain for the whole communication AFTER having authenticated

jhawkesworth commented:
Yeah, lookup makes sense. Callbacks really meant for reporting anyway, so it's bending things to use for Kerberos tickets. Forget that I commented!

azenk commented:
While reading the comment from @bcoca it occurred to me that lastpass would be a better name for this plugin, so I'll make that change and clean up some white space issues.

azenk commented:
I added some tests and refactored things a little.  It looks like all the tests are passing now.  I think it's ready to go.  I'm happy to make further improvements if requested though.

DebbieGillespie commented:
We would really like this feature. Could this pull request be merged?

sivel commented:
My 2 cents here.  I would much rather see this use the lastpass API, potentially through something like `lastpass-python` or even just direct API calls instead of shelling out to lastpass-cli.

azenk commented:
Are there docs available on that?  I looked for API docs prior to writing this and did not find any mention of anything other than the provisioning API.  This plugin has been working fine for us.  One nice benefit of using the cli, is that the lastpass vault session management is handled outside of ansible.   

Provided the session management could be dealt with properly (not using ansible vault to store my lastpass credentials) I'd be happy to modify this to use the API, if you can point me towards docs.

azenk commented:
As far as I can tell, lastpass-python cannot be used without providing your username and password at runtime, which would make the plugin useless for me.  So, I won't be using that.  Apart from shelling out to the cli, are there any other objections to merging this?

georgepsarakis commented:
@azenk there is a function for executing shell commands named [run_cmd](https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/cmd_functions.py#L27).

Perhaps it would be preferrable to use this function, instead of using `subprocess.Popen` directly?

georgepsarakis commented:
@azenk regarding the API, if I understand correctly from the code, both the CLI and the
`lastpass-python` library make certain HTTP requests that return XML.

I think it should be also noted that `lastpass-python` is not officially supported whereas the CLI is.

azenk commented:
@georgepsarakis I could swap run_cmd in, but I'm not sure I see an advantage other than maintaining consistency with other ansible code.

georgepsarakis commented:
@azenk yes that is the main reason. Reusing well-tested code in large projects is beneficial, don't you think?

azenk commented:
@georgepsarakis Yes, absolutely.  In this case it's slightly problematic since I'd have to build an escaped command string just to have shlex.split return the same list of args that I started with (is there a good way to do this in python 2.x?  3.3 has shlex.quote).  While this is possible, I think the advantages of skipping all that and using Popen directly outweigh the negatives.  It's certainly doable though.

azenk commented:
What's the status of this?  I'm not planning to make any revisions to this unless someone offers concrete feedback about changes that are required prior to merging.

azenk commented:
Is there any chance this could be merged for 2.3?

Dispader commented:
Very excited to have this new feature— anyone have any more thoughts, here?

mattclay commented on lib/ansible/plugins/lookup/lastpass.py:
```diff
@@ -0,0 +1,79 @@
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables=None, **kwargs):
+        lp = kwargs.get('__mock_lpass_obj', LPass())
```
Instead of making the plugin aware of the tests, use `mock.patch` to mock the `LPass` class in the tests.

mattclay reviewed (changes requested):
@azenk Looks good overall. Just one change to how unit tests are implemented should be made.

azenk committed:
Refactor to use mock.patch for testing

mattclay pointed out that using mock.patch would be superior to having
the lookup plugin be aware of the mock.

azenk commented:
@mattclay I think I've addressed the issue.  I'm a bit of a novice with mock.patch, so please forgive any rookie mistakes, etc.

mattclay commented on test/units/plugins/lookup/test_lastpass.py:
```diff
@@ -26,6 +26,8 @@
 from argparse import ArgumentParser
 import six
 
+from mock import patch
```
Sorry, I should have mentioned this in my original feedback. Import `patch` using:

```python
from ansible.compat.tests.mock import patch
```

mattclay reviewed (changes requested):
Could you amend your last commit message to remove the @ mention of me. That should help cut down on spurious notifications on GitHub.

Other than that, and the one requested change, it looks good. Thanks for the quick response to feedback!

azenk commented:
@mattclay  So amended.  I also rolled the import change into the amended commit.

mattclay commented:
@azenk Thanks for the new lookup plugin!
